docs/protocol.md

```diff
@@ -0,0 +1,230 @@
+
+The current RelayClient and RelayServer protocol (described in this document) is `0.4.1`
+
+The client should validate the relay supports this version when processing the `/getaddr` response (see below) 
+
+Version codes folow [SEMVER](https://semver.org)
+
+## Terminology
+
+* **client** - the calling client. identified by an address (And has the corresponding private-key or method for signing)
+* **relay** or **relayer** - the server that creates the actual ethereum transaction. 
+* **target** or **target contract** - the contract that receive the call and will pay for the transaction.
+* **RelayHub** - the on-chain RelayHub contract.
+* **relayed call** - the call to be sent over the blockchain. This is a signed request, (by the client) 
+    but it is not an ethereum transaction: it contains relaying-specific fields. 
+
+## Overview
+
+At a high-level the client perform the following steps:
+
+1. [Create list of Potential Relays](#create-list)
+2. [Select a relay](#select-relay)
+3. [Create a request](#create-request)
+4. [Send Request to the Relay.](#send-request)
+5. [Validate Response.](#validate-response)
+6. [ Handle Relay Error Responses.](#relay-error)
+7. [Process Trasnaction-receipt](#process-tx-rcpt)
+
+A relay perform the following steps:
+
+1. [Relay setup and background processes](#relay-setup)
+2. [Handling Incoming Request](#relay-process)
+3. [Handling stuck transactions](#relay-underpriced-tx)
+
+## Client Detailed Request Flow
+<a name="create-list"></a> 
+### 1. Create a List of Potential Relays
+
+* First, the client MUST get the RelayHub used by the target contract, by calling `target.getHubAddr()`<br/>
+    **NOTE**: it is important that the client **doesn't assume** the target uses the global RelayHub address: If/when the RelayHub is updated, its the target contract's responsibility to update its hub link.
+* Next, the client SHOULD filter for recent `RelayAdded` and `RelayRemoved` events sent by the RelayHub.
+  Since a relay is required to send such event every 6000 blocks (roughly 24hours), the client should look at most at the latest 6000 blocks.     
+* For each relay, look for the latest `RelayAdded` event. Also, all Relays with `RelayRemoved` event should be removed from the list.
+* Now the client should filter out relays that it doesn't care about. e.g.: 
+  Ignore relays with stake time or stake delay below a given threshold. 
+* Sort the relays in the preferred order. e.g.: prefer relays with lower transaction fee, and also depend on the [Relay Reputation](#relay-reputation)
+* Dynamic relay selection: before making the call, the client SHOULD "ping" the relay (see below)
+* For each relay, the client keeps the relay `address` and `url`  
+
+<a name="select-relay"></a>
+###  2. Select a Relay
+
+* For each potential relay, the client "pings" the relay by sending a `/getaddr` request.
+* Validate the relay is valid (contains `valid:true`)
+* Validate the relay supports this protocol: `version:1.0.x`
```

Comment:
Is this advertised by the relayer?

Reply:
tricky: the relay reports the *relay protocol* version, which is currently `0.4.1`
the relay-client protocol is not directly related to the RelayHub protocol, but it is heavily influenced by it (as  the client signs a struct that is checked by the RelayHub)